src/ssim.jl

```diff
@@ -0,0 +1,121 @@
+"""
+    SSIM([kernel], [(α, β, γ)]) <: FullReferenceIQI
+    assess(iqi::SSIM, img, ref)
+    ssim(img, ref)
+
+Structural similarity (SSIM) index is an image quality assessment method based
+on degradation of structural information.
+
+The SSIM index is composed of three components: luminance similarity, contrast
+similarity, and structural similarity. `ssim = 𝐿ᵅ * 𝐶ᵝ * 𝑆ᵞ`,
+where `W := (α, β, γ)` controls relative importance of each components.
+By default `W = (1.0, 1.0, 1.0)`.
+
+In practice, a local version SSIM is used. At each pixel, SSIM is calculated
+locally with neighborhoods weighted by guassian `kernel`, returning a ssim map;
+`ssim` is actaully `mean(ssim_map)`.
+By default `kernel = KernelFactors.gaussian(1.5, 11)`.
+
+!!! info
+
+    RGB images are treated as 3d Gray images, general Color3 images are
+    converted to RGB images first.
+
+# Example
+
+In most cases, `ssim(img, ref)` should be sufficient, which calls the default
+SSIM. One could also instantiate a SSIM and pass it to `assess`, for example:
+
+```julia
+iqi = SSIM(KernelFactors.gaussian(2.5, 17), (1.0, 1.0, 2.0))
+assess(iqi, img, ref)
+```
+
+# Reference
+
+[1] Wang, Z., Bovik, A. C., Sheikh, H. R., & Simoncelli, E. P. (2004). Image quality assessment: from error visibility to structural similarity. _IEEE transactions on image processing_, 13(4), 600-612.
+
+[2] Wang, Z., Bovik, A. C., Sheikh, H. R., & Simoncelli, E. P. (2003). The SSIM Index for Image Quality Assessment. Retrived May 30, 2019, from http://www.cns.nyu.edu/~lcv/ssim/
+"""
+struct SSIM <: FullReferenceIQI
+    kernel::OffsetArray
+    W::NTuple{3}
+    function SSIM(kernel, W)
+        ndims(kernel) == 1 || throw(ArgumentError("only 1-d kernel is valid"))
+        all(W .>= 0) || throw(ArgumentError("W should be non-negative"))
+        new(centered(kernel), W)
+    end
+end
+
+# default values from [1]
+const SSIM_KERNEL = KernelFactors.gaussian(1.5, 11) # kernel
+const SSIM_W = (1.0, 1.0, 1.0) # (α, β, γ)
+SSIM(kernel=SSIM_KERNEL) = SSIM(kernel, SSIM_W)
+
+# api
+(iqi::SSIM)(x, ref) = mean(_ssim_map(iqi, x, ref))
```

Comment:
Instead of `mean(_ssim_map(iqi, x, ref))`, [2] drops its padding edge, i.e., `ssim = mean(_ssim_map[6:end-5, 6:end-5])`

I did some numerical tests on
`img1`: lean_gray_512.tif
`img2`: cameraman.tif

and there're three results returned:

* `[2]`: `0.3526`
* [tensorflow](https://www.tensorflow.org/api_docs/python/tf/image/ssim): `0.35255808`

.

I guess it would be okay to follow MATLAB built-in results.

* this PR: `0.3594783288625984`
* MATLAB built-in ssim: `0.3595`
* [pyssim](https://github.com/jterrace/pyssim): `0.3594754`

.
* scikit-image: `0.35229376382502575` <-- `measure.compare_ssim(img1, img2, gaussian_weights=True, sigma=1.5, use_sample_covariance=False)`